File: apps/web/lib/sportsbook-odds-input.ts
export const SPORTSBOOK_ODDS_FORMAT_MESSAGE =
  "Enter decimal odds using a full stop, for example 8.5.";
export const SPORTSBOOK_ODDS_MINIMUM_MESSAGE = "Enter odds of 1.01 or higher.";

const completeDecimalInputPattern = /^[0-9]+(?:\.[0-9]+)?$/;
const simpleDecimalCommaPattern = /^[0-9]+,[0-9]{1,2}$/;
const fractionalOddsPattern = /^([0-9]+)\/([0-9]+)$/;

export function hasCompleteDecimalInputSyntax(value: string): boolean {
  return completeDecimalInputPattern.test(value);
}

export function getSportsbookOddsInputError(
  value: string,
  options: { required?: boolean } = {}
): string | null {
  if (value === "") {
    return options.required ? "Enter odds." : null;
  }
  if (!hasCompleteDecimalInputSyntax(value)) {
    return SPORTSBOOK_ODDS_FORMAT_MESSAGE;
  }
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) {
    return SPORTSBOOK_ODDS_FORMAT_MESSAGE;
  }
  if (parsed < 1.01) {
    return SPORTSBOOK_ODDS_MINIMUM_MESSAGE;
  }
  return null;
}

export function parseSportsbookOddsInput(value: string): number | null {
  if (getSportsbookOddsInputError(value) !== null || value === "") {
    return null;
  }
  return Number(value);
}

export type CalculatorOddsNormalization = {
  canonicalValue: string;
  converted: boolean;
  error: string | null;
};

export function normalizeCalculatorOddsInput(value: string): CalculatorOddsNormalization {
  let canonicalValue = value;
  if (simpleDecimalCommaPattern.test(value)) {
    canonicalValue = value.replace(",", ".");
  } else {
    const fraction = fractionalOddsPattern.exec(value);
    if (fraction) {
      const numerator = BigInt(fraction[1]);
      const denominator = BigInt(fraction[2]);
      if (denominator === BigInt(0)) {
        return { canonicalValue: value, converted: false, error: "Fractional odds need a denominator above zero." };
      }
      const hundred = BigInt(100);
      const scaledNumerator = (numerator + denominator) * hundred;
      const roundedHundredths = (scaledNumerator / denominator) +
        ((scaledNumerator % denominator) * BigInt(2) >= denominator ? BigInt(1) : BigInt(0));
      canonicalValue = `${roundedHundredths / hundred}.${String(roundedHundredths % hundred).padStart(2, "0")}`;
    }
  }
  const error = getSportsbookOddsInputError(canonicalValue, { required: true });
  return { canonicalValue, converted: canonicalValue !== value && error === null, error };
}

File: apps/api/src/openforge_api/sportsbook_odds_input.py
from __future__ import annotations

import json
import re
from decimal import ROUND_HALF_UP, Decimal
from typing import Any

from pydantic_core import PydanticCustomError

SPORTSBOOK_ODDS_FORMAT_MESSAGE = "Enter decimal odds using a full stop, for example 8.5."
SPORTSBOOK_ODDS_MINIMUM_MESSAGE = "Enter odds of 1.01 or higher."
SPORTSBOOK_ODDS_MINIMUM = Decimal("1.01")

_DECIMAL_ODDS_PATTERN = re.compile(r"^[0-9]+(?:\.[0-9]+)?$")
_SIMPLE_DECIMAL_COMMA_PATTERN = re.compile(r"^[0-9]+,[0-9]{1,2}$")
_FRACTIONAL_ODDS_PATTERN = re.compile(r"^([0-9]+)/([0-9]+)$")
_NESTED_ODDS_KEYS = ("layOdds", "placedLayOdds")


def validate_complete_decimal_string(value: Any, *, message: str) -> str:
    """Validate an unsigned base-10 input without trimming or numeric coercion."""

    if not isinstance(value, str) or not _DECIMAL_ODDS_PATTERN.fullmatch(value):
        raise PydanticCustomError("decimal_input_format", message)
    return value


def validate_sportsbook_odds(value: Any, *, allow_empty: bool = True) -> str:
    """Validate the complete request value without normalising or repairing it."""

    if not isinstance(value, str):
        raise PydanticCustomError("sportsbook_odds_format", SPORTSBOOK_ODDS_FORMAT_MESSAGE)
    if value == "":
        if allow_empty:
            return value
        raise PydanticCustomError("sportsbook_odds_required", "Odds are required.")
    try:
        validate_complete_decimal_string(value, message=SPORTSBOOK_ODDS_FORMAT_MESSAGE)
    except PydanticCustomError as exc:
        raise PydanticCustomError("sportsbook_odds_format", SPORTSBOOK_ODDS_FORMAT_MESSAGE) from exc

    decimal_value = Decimal(value)
    if not decimal_value.is_finite():
        raise PydanticCustomError("sportsbook_odds_format", SPORTSBOOK_ODDS_FORMAT_MESSAGE)
    if decimal_value < SPORTSBOOK_ODDS_MINIMUM:
        raise PydanticCustomError("sportsbook_odds_minimum", SPORTSBOOK_ODDS_MINIMUM_MESSAGE)
    return value


def normalize_calculator_odds(value: Any) -> str:
    """Return canonical decimal calculator odds without accepting ambiguous syntax.

    Fractional entry is converted to the existing two-decimal odds display contract.
    The calculation itself uses Decimal throughout; ordinary decimal input is not
    rounded or otherwise repaired.
    """

    if not isinstance(value, str):
        raise PydanticCustomError("calculator_odds_format", SPORTSBOOK_ODDS_FORMAT_MESSAGE)
    normalized = value
    comma_match = _SIMPLE_DECIMAL_COMMA_PATTERN.fullmatch(value)
    fraction_match = _FRACTIONAL_ODDS_PATTERN.fullmatch(value)
    if comma_match:
        normalized = value.replace(",", ".")
    elif fraction_match:
        numerator = Decimal(fraction_match.group(1))
        denominator = Decimal(fraction_match.group(2))
        if denominator == 0:
            raise PydanticCustomError(
                "calculator_odds_zero_denominator", "Fractional odds need a denominator above zero."
            )
        decimal_odds = (Decimal("1") + (numerator / denominator)).quantize(
            Decimal("0.01"), rounding=ROUND_HALF_UP
        )
        normalized = f"{decimal_odds:.2f}"
    return validate_sportsbook_odds(normalized, allow_empty=False)


def validate_nested_sportsbook_odds(serialized: Any) -> str:
    if not isinstance(serialized, str):
        raise PydanticCustomError(
            "sportsbook_nested_odds_format", "Multi-lay outcomes must be a JSON string."
        )
    try:
        entries = json.loads(serialized or "[]")
    except json.JSONDecodeError as exc:
        raise PydanticCustomError(
            "sportsbook_nested_odds_format", "Multi-lay outcomes must contain valid JSON."
        ) from exc
    if not isinstance(entries, list):
        raise PydanticCustomError(
            "sportsbook_nested_odds_format", "Multi-lay outcomes must be a JSON array."
        )

    for index, entry in enumerate(entries):
        if not isinstance(entry, dict):
            continue
        for key in _NESTED_ODDS_KEYS:
            if key not in entry or entry[key] == "":
                continue
            try:
                validate_sportsbook_odds(entry[key])
            except PydanticCustomError as exc:
                raise PydanticCustomError(
                    "sportsbook_nested_odds_value",
                    f"multi_lay_outcomes_json[{index}].{key}: {exc.message_template}",
                ) from exc
    return serialized
